diff --git a/configure.ac b/configure.ac
index 1e05aa3..74f3920 100644
--- a/configure.ac
+++ b/configure.ac
@@ -134,6 +134,24 @@ fi
 AC_SUBST(JSON_GLIB_CFLAGS)
 AC_SUBST(JSON_GLIB_LIBS)
 
+
+dnl ******************************
+dnl
+dnl gtk-mac-integration
+dnl
+PKG_CHECK_MODULES(MAC_INTEGRATION,
+		  [gtk-mac-integration],
+		  [enable_mac_integration=yes],
+		  [enable_mac_integration=no])
+
+if test x$enable_mac_integration = xyes ; then
+	AC_DEFINE(HAVE_MAC_INTEGRATION, 1, [Define to 1 if gtk-mac-integration support is included])
+fi
+
+AC_SUBST(MAC_INTEGRATION_CFLAGS)
+AC_SUBST(MAC_INTEGRATION_LIBS)
+
+
 AM_CONDITIONAL(ENABLE_JSON_GLIB, test x"$enable_json_glib" != x"no")
 
 dnl ******************************
@@ -247,7 +265,7 @@ dnl ===========================================================================
 dnl
 dnl SM Client
 dnl
-GDK_TARGET="$($PKG_CONFIG --variable target gdk-2.0)"
+GDK_TARGET="$($PKG_CONFIG --variable target gdk-3.0)"
 
 AC_MSG_CHECKING([which smclient backend to use])
 AC_ARG_WITH([smclient],
@@ -365,5 +383,5 @@ Configuration:
 	Use libmagic            : ${enable_magic}
 	JSON support            : ${enable_json_glib}
 	Notifications support   : ${enable_notification}
-	Use libarchive          : ${enable_libarchive}
+	Use libarchive          : ${enable_libarchive},${LIBARCHIVE_LIBS}
 "
diff --git a/copy-n-paste/Makefile.am b/copy-n-paste/Makefile.am
index 59789f4..efd4a54 100644
--- a/copy-n-paste/Makefile.am
+++ b/copy-n-paste/Makefile.am
@@ -4,7 +4,7 @@ AM_CPPFLAGS = $(DISABLE_DEPRECATED) \
 noinst_LTLIBRARIES = libeggsmclient.la
 
 libeggsmclient_la_LIBADD = $(GTK_LIBS)
-libeggsmclient_la_CFLAGS = $(GTK_CFLAGS)
+libeggsmclient_la_CFLAGS = $(GTK_CFLAGS) $(MAC_INTEGRATION_CFLAGS)
 
 libeggsmclient_la_SOURCES = 			\
 	eggdesktopfile.h 			\
diff --git a/copy-n-paste/eggdesktopfile.c b/copy-n-paste/eggdesktopfile.c
index e6692e1..0c4fff3 100644
--- a/copy-n-paste/eggdesktopfile.c
+++ b/copy-n-paste/eggdesktopfile.c
@@ -31,7 +31,11 @@
 #include <unistd.h>
 
 #include <glib/gi18n.h>
+
+#ifndef MAC_INTEGRATION
 #include <gdk/gdkx.h>
+#endif
+
 #include <gtk/gtk.h>
 
 struct EggDesktopFile {
diff --git a/src/Makefile.am b/src/Makefile.am
index dfc0958..4691381 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -43,7 +43,8 @@ AM_CPPFLAGS =						\
 	$(FR_CFLAGS)					\
 	$(JSON_GLIB_CFLAGS)				\
 	$(LIBNOTIFY_CFLAGS)				\
-	$(LIBARCHIVE_CFLAGS)
+	$(LIBARCHIVE_CFLAGS)			\
+	$(MAC_INTEGRATION_CFLAGS)
 
 BUILT_SOURCES =			\
 	fr-marshal.c		\
@@ -226,7 +227,8 @@ file_roller_LDADD =					\
 	$(FR_LIBS)					\
 	$(JSON_GLIB_LIBS)				\
 	$(LIBNOTIFY_LIBS)				\
-	$(LIBARCHIVE_LIBS)
+	$(LIBARCHIVE_LIBS)				\
+	$(MAC_INTEGRATION_LIBS)
 
 if ENABLE_MAGIC
 file_roller_LDADD += $(MAGIC_LIBS)
diff --git a/src/dlg-package-installer.c b/src/dlg-package-installer.c
index 6fa8b4c..c925991 100644
--- a/src/dlg-package-installer.c
+++ b/src/dlg-package-installer.c
@@ -22,7 +22,11 @@
 #include <config.h>
 #include <string.h>
 #include <glib/gi18n.h>
+
+#ifndef MAC_INTEGRATION
 #include <gdk/gdkx.h>
+#endif
+
 #include <gtk/gtk.h>
 #include "dlg-package-installer.h"
 #include "gio-utils.h"
diff --git a/src/fr-application.c b/src/fr-application.c
index 60ea595..f1e222a 100644
--- a/src/fr-application.c
+++ b/src/fr-application.c
@@ -41,6 +41,9 @@
 #include "glib-utils.h"
 #include "gtk-utils.h"
 
+#ifdef MAC_INTEGRATION
+#include<gtkosxapplication.h>
+#endif
 
 #define ORG_GNOME_ARCHIVEMANAGER_XML "/org/gnome/FileRoller/../data/org.gnome.ArchiveManager1.xml"
 #define SERVICE_TIMEOUT 10
@@ -453,6 +456,10 @@ struct _FrApplicationPrivate {
 	guint          owner_id;
 	GSettings     *listing_settings;
 	GSettings     *ui_settings;
+    
+#ifdef MAC_INTEGRATION
+    GtkosxApplication *theApp;
+#endif
 };
 
 
@@ -467,7 +474,7 @@ fr_application_finalize (GObject *object)
 		g_bus_unown_name (self->priv->owner_id);
 	_g_object_unref (self->priv->listing_settings);
 	_g_object_unref (self->priv->ui_settings);
-
+    _g_object_unref (self->priv->theApp);
 	release_data ();
 
         G_OBJECT_CLASS (fr_application_parent_class)->finalize (object);
@@ -651,7 +658,7 @@ fr_application_command_line (GApplication            *application,
 
 	if (remaining_args == NULL) { /* No archive specified. */
 		if (! arg_service)
-			gtk_widget_show (fr_window_new ());
+			gtk_window_present (GTK_WINDOW(fr_window_new ()));
 		return fr_application_command_line_finished (application, EXIT_SUCCESS);
 	}
 
@@ -732,7 +739,7 @@ fr_application_command_line (GApplication            *application,
 			GFile     *file;
 
 			window = fr_window_new ();
-			gtk_widget_show (window);
+			gtk_window_present (GTK_WINDOW(window));
 
 			file = g_application_command_line_create_file_for_arg (command_line, filename);
 			fr_window_archive_open (FR_WINDOW (window), file, GTK_WINDOW (window));
@@ -822,6 +829,32 @@ fr_application_class_init (FrApplicationClass *klass)
 	application_class->activate = fr_application_activate;
 }
 
+static void
+fr_application_open_files(FrApplication *self,gchar *path,GtkosxApplication *theApp)
+{
+    GtkWidget *window = NULL;
+    GFile *file;
+    
+    
+    GList *link;
+    
+	for (link = gtk_application_get_windows (GTK_APPLICATION (self));
+	     link != NULL;
+	     link = link->next)
+	{
+        if(fr_window_get_archive_file(FR_WINDOW (link->data)) == NULL){
+            window = GTK_WIDGET (link->data);
+            break;
+        }
+	}
+    
+    if(window == NULL)window = fr_window_new ();
+    gtk_window_present(GTK_WINDOW(window));
+    file = g_file_new_for_path (path);
+    fr_window_archive_open (FR_WINDOW (window), file, GTK_WINDOW (window));
+    
+    g_object_unref (file);
+}
 
 static void
 fr_application_init (FrApplication *self)
@@ -844,6 +877,10 @@ fr_application_init (FrApplication *self)
 	self->priv->introspection_data = NULL;
 	self->priv->listing_settings = g_settings_new (FILE_ROLLER_SCHEMA_LISTING);
 	self->priv->ui_settings = g_settings_new (FILE_ROLLER_SCHEMA_UI);
+    self->priv->theApp = g_object_new(GTKOSX_TYPE_APPLICATION, NULL);
+    
+    g_signal_connect_swapped(self->priv->theApp,"NSApplicationOpenFile",G_CALLBACK(fr_application_open_files),self);
+    gtkosx_application_ready (self->priv->theApp);
 }
 
 
diff --git a/src/fr-command-lrzip.c b/src/fr-command-lrzip.c
index ad53a13..6bba7ba 100644
--- a/src/fr-command-lrzip.c
+++ b/src/fr-command-lrzip.c
@@ -45,7 +45,11 @@ list__process_line (char     *line,
 
 	struct stat st;
 	if (stat (comm->filename, &st) == 0)
+#ifdef __APPLE__		
+		fdata->modified = st.st_mtimespec.tv_sec;
+#else
 		fdata->modified = st.st_mtim.tv_sec;
+#endif
 	else
 		time(&(fdata->modified));
 
diff --git a/src/fr-init.c b/src/fr-init.c
index 2c3ad75..e449816 100644
--- a/src/fr-init.c
+++ b/src/fr-init.c
@@ -57,6 +57,9 @@
 #include "typedefs.h"
 #include "preferences.h"
 
+#ifdef MAC_INTEGRATION
+#include <gtkosxapplication.h>
+#endif
 
 /* The capabilities are computed automatically in
  * compute_supported_archive_types() so it's correct to initialize to 0 here. */
@@ -171,6 +174,11 @@ FrExtensionType file_ext_type[] = {
 };
 
 
+const char*just4icon[]={
+	"gnome-mime-application-x-archive",
+	"package-x-generic"
+};
+
 GList        *CommandList;
 gint          ForceDirectoryCreation;
 GHashTable   *ProgramsCache;
@@ -623,10 +631,18 @@ initialize_data (void)
 					       g_str_equal,
 					       g_free,
 					       NULL);
+#ifdef MAC_INTEGRATION
+	gchar *res_path = gtkosx_application_get_resource_path();
+	if(res_path){
+		gchar *icon_path = g_build_filename(res_path,"/share/"PACKAGE_NAME"/icons",NULL);
+		gtk_icon_theme_append_search_path(gtk_icon_theme_get_default (),icon_path);
+		g_free(res_path);
+		g_free(icon_path);
+	}
+#endif
 
 	gtk_icon_theme_append_search_path (gtk_icon_theme_get_default (),
 					   PKG_DATA_DIR G_DIR_SEPARATOR_S "icons");
-
 	migrate_options_directory ();
 	register_archives ();
 	compute_supported_archive_types ();
diff --git a/src/ui/app-menu.ui b/src/ui/app-menu.ui
index cd43381..ca25172 100644
--- a/src/ui/app-menu.ui
+++ b/src/ui/app-menu.ui
@@ -1,9 +1,16 @@
 <interface>
   <menu id="app-menu">
+  	<section>
+	    <item>
+	      <attribute name="action">app.about</attribute>
+	      <attribute name="label" translatable="yes">_About Archive Manager</attribute>
+	    </item>	
+	</section>
     <section>
       <item>
         <attribute name="action">app.new</attribute>
         <attribute name="label" translatable="yes">_New Archive</attribute>
+		<attribute name="accel"><![CDATA[<Primary>N]]></attribute>
       </item>
     </section>
     <section>
@@ -26,17 +33,14 @@
       </item>
     </section>
     <section>
-      <item>
+      <!--item>
         <attribute name="action">app.help</attribute>
         <attribute name="label" translatable="yes">_Help</attribute>
-      </item>
-      <item>
-        <attribute name="action">app.about</attribute>
-        <attribute name="label" translatable="yes">_About Archive Manager</attribute>
-      </item>
+      </item -->
       <item>
         <attribute name="action">app.quit</attribute>
         <attribute name="label" translatable="yes">_Quit</attribute>
+		<attribute name="accel"><![CDATA[<Primary>Q]]></attribute>
       </item>
     </section>
   </menu>
